File: C/1304/B/solution.c
// Author: Shivshankar Pirgonde
// Adapted from online resources and rewritten for learning.

#include <stdio.h>
#include <string.h>

#define MAXN 105 // Max no. of input strings.
#define MAXM 55  // Max length of each input string.

// Function Prototypes
void reverseString(char str[], int len);
void checkReverse(const char s1[], const char s2[], int m);
void solve(int n, int m, char inputString[MAXN][MAXM]);

int main()
{

    // n = No. of strings entered.
    // m = Exact length of each entered string .

    int n, m;

    scanf("%d %d", &n, &m);

    char inputString[MAXN][MAXM];

    for (int i = 0; i < n; i++)
    {
        scanf("%s", inputString[i]);
    }

    solve(n, m, inputString);

    return 0;
}

// Helper Funtion to reverse a string (Pass by Reference)
void reverseString(char str[], int len)
{
    for (int l = 0, r = len - 1; l < r; l++, r--)
    {
        char temp = str[l];
        str[l] = str[r];
        str[r] = temp;
    }
}

// check if two strings are reverse of each other
int isReverse(const char s1[], const char s2[], int m)
{
    for (int i = 0; i < m; i++)
    {
        if (s1[i] != s2[m - 1 - i])
            return 0;
    }
    return 1;
}

// check if string is a pallindrome
int isPallindrome(const char s[], int m)
{
    for (int l = 0, r = m - 1; l < r; l++, r--)
    {
        if (s[l] != s[r])
            return 0;
    }
    return 1;
}

// Function to solve the problem
void solve(int n, int m, char inputString[MAXN][MAXM])
{
    int used[MAXN] = {0};
    int leftIndices[MAXN], leftCount = 0;
    int middleIndex = -1;

    // 1.Identify reverse pairs and middle pallindrome
    for (int i = 0; i < n; i++)
    {
        if (used[i])
            continue;

        // Try to find matching reverse pair for the left side
        int foundPair = 0;
        for (int j = i + 1; j < n; j++)
        {
            if (!used[j] && isReverse(inputString[i], inputString[j], m))
            {
                leftIndices[leftCount++] = i;
                used[i] = used[j] = 1;
                foundPair = 1;
                break;
            }
        }

        // If no reverse pair, check if it's a candidate for the middle
        if (!foundPair && middleIndex == -1 && isPallindrome(inputString[i], m))
        {
            middleIndex = i;
            used[i] = 1;
        }
    }

    // 2.Output the result directly

    int totalLen = (leftCount * 2 * m) + (middleIndex != -1 ? m : 0);
    printf("%d\n", totalLen);

    for (int i = 0; i < leftCount; i++)
        printf("%s", inputString[leftIndices[i]]);

    if (middleIndex != -1)
        printf("%s", inputString[middleIndex]);

    for (int i = leftCount - 1; i >= 0; i--)
    {
        char temp[MAXM];
        strcpy(temp, inputString[leftIndices[i]]);
        reverseString(temp, m);
        printf("%s", temp);
    }

    printf("\n");
}

File: Go/1304/B/solution.go
// Author: Shivshankar Pirgonde

package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {

	//Fast I/O setup

	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)

	if !scanner.Scan() {
		return
	}

	//Read n and m
	var n, m int
	fmt.Sscanf(scanner.Text(), "%d", &n)
	if !scanner.Scan() {
		return
	}
	fmt.Sscanf(scanner.Text(), "%d", &m)

	//Pre-allocate the slice capacity to avoid resizing
	inputStrings := make([]string, 0, n)
	for i := 0; i < n; i++ {
		if scanner.Scan() {
			inputStrings = append(inputStrings, scanner.Text())
		}
	}

	solve(n, m, inputStrings)
}

//Function to solve the problem

func solve(n, m int, inputStrings []string) {
	counts := make(map[string]int)
	for _, s := range inputStrings {
		counts[s]++
	}

	var leftPart []string
	var middlePart string

	//1.Iterate through the original list to maintain a consistent order

	for _, s := range inputStrings {
		rev := reverse(s)

		if s == rev {
			//case 1: self palindromes like 'aba'
			//If we have at least 2 they can be pair, one on left other on right
			if counts[s] >= 2 {
				leftPart = append(leftPart, s)
				counts[s] = counts[s] - 2
			}

			//If 1 remains, it can be single middle piece
			if middlePart == "" && counts[s] == 1 {
				middlePart = s
				counts[s]--
			}
		} else {
			//case 2: Reverse Pairs like 'abc' and 'cba'
			if counts[s] > 0 && counts[rev] > 0 {
				leftPart = append(leftPart, s)
				counts[s]--
				counts[rev]--
			}

		}
	}

	//2.Assebmle the final string

	//Exact Preallocation
	totalLen := (len(leftPart) * 2 * m)
	if middlePart != "" {
		totalLen = totalLen + m
	}

	var builder strings.Builder

	//Grow reserves exactly the bytes we need in one go
	builder.Grow(totalLen)

	//3.Asseble complete string
	for _, s := range leftPart {
		builder.WriteString(s)
	}

	builder.WriteString(middlePart)

	for i := len(leftPart) - 1; i >= 0; i-- {
		builder.WriteString(reverse(leftPart[i]))
	}

	//4.Output
	result := builder.String()

	fmt.Println(len(result))
	fmt.Println(result)

}

// Helper func to reverse a string
func reverse(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}
